### PR TITLE
Read args from env as an alternative, Dockerize vj4

### PR DESCRIPTION
**Pull request by moesoha**
~~And why not accept arguments from both command-line and environment variables?~~

We can use environment variables to pass arguments now. For example, `python3 -m vj4.server --db-name vijos` equals `VJ_DB_NAME=vijos python3 -m vj4.server`.

### CODE REVIEW

**Comment by breeswish** [comment]
@moesoha It is just not implemented. Feel free to add support.

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+WORKDIR /app/src
+RUN npm install
+RUN npm run build:production
+
+# main
+FROM python:3.6-stretch
+COPY --from=stage-node /app/src /app/src
+WORKDIR /app/src
+
+RUN apt-get update \
```
Comment (twd2) [inline Dockerfile]: Could you please use just one `RUN` like https://github.com/vijos/jd4/blob/master/Dockerfile#L3 ?

File: docker-compose.yml [hunk 2]
```diff
@@ -0,0 +1,27 @@
+version: '3'
+services:
+  web:
+    build: .
+    image: vijos/vj4
+    restart: always
+    # set arguments for vj4.server below
+    command: vj4.server --db-host=mongodb --mq-host=rabbitmq --listen=http://0.0.0.0:80
```
Comment (twd2) [inline docker-compose.yml]: `listen=http://0.0.0.0:80` might be a bad default setting.
What about unix socket?

File: .env.example [hunk 3]
```diff
@@ -0,0 +1,173 @@
+# # # # # # # # # # # # # # # # # # #
```
Comment (twd2) [inline .env.example]: Is this file automatically generated from the source code?

File: vj4/util/options.py [hunk 4]
```diff
@@ -21,7 +21,15 @@ def define(self, name, default=None, help=None):
 
   def __getattr__(self, item):
     if self._dirty:
-      self._parser.parse_known_args(args=sys.argv[1:], namespace=self._namespace)
+      args_to_parse = []
```
Comment (twd2) [inline vj4/util/options.py]: @iceb0y PTAL

File: .env.example (hunk 3 above)
Comment (moesoha) [reply .env.example]: @twd2 It seems that is not easy to generate .env from codes. We cannot get such as name, help and default value of arguments from ArgumentParser.

Arguments won't change often and many software write examples for `.env` manually as well, so I suggest just keeping this.

File: vj4/server.py [hunk 5]
```diff
@@ -15,6 +16,9 @@
 options.define('listen', default='http://127.0.0.1:8888', help='Server listening address.')
 options.define('prefork', default=1, help='Number of prefork workers.')
 options.define('syslog', default=False, help='Use syslog instead of stderr for logging.')
+options.define('listen_owner', default='www-data', help='Owner of the socket when server is listening to a unix socket.')
```
Comment (twd2) [inline vj4/server.py]: These changes can be extracted to a new PR so that I can merge these instantly.

File: vj4/server.py [hunk 6]
```diff
@@ -41,6 +45,8 @@ def main():
     except FileNotFoundError:
       pass
     sock.bind(url.path)
+    shutil.chown(url.path, options.listen_owner, options.listen_group)
```
Comment (twd2) [inline vj4/server.py]: ditto

File: vj4/server.py (hunk 5 above)
Comment (moesoha) [reply vj4/server.py]: OK, i'm going to cherry-pick this commit and create a new pr.

File: .env.example (hunk 3 above)
Comment (twd2) [reply .env.example]: Synchronizing between the source code and this file requires more developers' effort.

But I'm lazy. ^\_^ T\_T

File: vj4/util/options.py [hunk 7]
```diff
@@ -21,7 +21,19 @@ def define(self, name, default=None, help=None):
...
+      for k, v in os.environ.items(): # using environments start with `VJ_` as arguments
+        if k.startswith("VJ_"):
+          if k == 'VJ_CMDLINE_FLAGS': # for setting bool flags like `--debug` or `--no-pretty`
+            for v in v.split(' '):
```
Comment (twd2) [inline vj4/util/options.py]: `args_to_parse.extend(v.split())`?

File: vj4/util/options.py [hunk 8]
```diff
@@ -21,7 +21,19 @@ def define(self, name, default=None, help=None):
 
   def __getattr__(self, item):
     if self._dirty:
-      self._parser.parse_known_args(args=sys.argv[1:], namespace=self._namespace)
+      args_to_parse = []
+      for k, v in os.environ.items(): # using environments start with `VJ_` as arguments
+        if k.startswith("VJ_"):
```
Comment (twd2) [inline vj4/util/options.py]: `'`

File: vj4/util/options.py [hunk 9]
```diff
@@ -21,7 +21,19 @@ def define(self, name, default=None, help=None):
...
+          if k == 'VJ_CMDLINE_FLAGS': # for setting bool flags like `--debug` or `--no-pretty`
+            for v in v.split(' '):
+              args_to_parse.append(v)
+          else:
+            args_to_parse.append('--' + k[3:].lower().replace('_','-'))
+            args_to_parse.append(v)
+      for v in sys.argv[1:]:
+        args_to_parse.append(v) # cmdline args can override one in env
```
Comment (twd2) [inline vj4/util/options.py]: `args_to_parse.extend(sys.argv[1:])`?

File: vj4/util/options.py (hunk 7 above)
Comment (moesoha) [reply vj4/util/options.py]: 丢人了丢人了我不知道有个extend（捂脸

File: .env.example (hunk 3 above)
Comment (moesoha) [reply .env.example]: @twd2 To be honest, I think command instructions here is not so useful. Just let them know the relationship between command-line arguments and environment variables, add prefix `VJ_` and replace `_` to `-`, is enough. The instruction here is just for a quicker start, just a few commonly used arguments is fine.

Hmm... About synchronizing, I believe the argument list won't be changed often.

**Comment by iceb0y** [comment]
Sorry for jumping in late, but have you considered to create a shell script to convert environment variables into command line arguments? The script can be like:

```bash
python -m vj4.server \
    --listen=${VJ4_LISTEN:-http://127.0.0.1:8888} \
    ...
```

For dockerize, I'm against putting it in this repository. Please create a separate repository for it. If you like I can create one under the `vijos` group. The script to support environment variables should be part of the docker repository and not here, but feel free to put one under the `examples/` directory.

Adding `listen_owner`, `listen_group` and `listen_mode` seems fine to me. But please make the default to do nothing (you changed the default behaviour in this PR!) to maintain backward compatibility.

Thanks!

**Comment by moesoha** [comment]
@iceb0y I've considered processing environment variables in shell. This form has less flexible than processing directly in codes, and may bring some trouble. I hope the docker image can be a out-of-the-box one, so I wrote the codes in `util/options.py`.

Create a repo in `vijos` is fine. I'll push files about dockerize to there after created.

**Comment by iceb0y** [comment]
moesoha ~ Created https://github.com/vijos/vj4-docker for you.

Using a shell script explicitly answers the question that, when environment variable and command line argument conflicts, which one is used. People can understand only by looking at the script and nothing (documentation or lower-level code) else.

If shell script doesn't provide enough flexibility for your use case, you could create a python script :)

**Comment by moesoha** [comment]
ok, thx.

Considering writing a python script as docker-entrypoint.

**Comment by twd2** [comment]
Moved to https://github.com/vijos/vj4-docker, closing.

**Pull request closed without merging**